inference/enterprise/workflows/execution_engine/executor/core.py

```diff
@@ -0,0 +1,163 @@
+from inference.enterprise.workflows.execution_engine.compiler.entities import (
+    CompiledWorkflow,
+)
+from inference.enterprise.workflows.execution_engine.compiler.utils import (
+    get_last_chunk_of_selector,
+)
+from inference.enterprise.workflows.execution_engine.executor.execution_cache import (
+    ExecutionCache,
+)
+from inference.enterprise.workflows.execution_engine.executor.flow_coordinator import (
+    ParallelStepExecutionCoordinator,
+    handle_flow_control,
+)
+from inference.enterprise.workflows.execution_engine.executor.output_constructor import (
+    construct_workflow_output,
+)
+from inference.enterprise.workflows.execution_engine.executor.parameters_assembler import (
+    assembly_step_parameters,
+)
+from inference_sdk.http.utils.iterables import make_batches
+
+
+async def run_workflow(
+    workflow: CompiledWorkflow,
+    runtime_parameters: Dict[str, Any],
+    max_concurrent_steps: int,
+    step_execution_mode: StepExecutionMode,
+) -> Dict[str, List[Any]]:
+    execution_cache = ExecutionCache.init()
+    execution_coordinator = ParallelStepExecutionCoordinator.init(
+        execution_graph=workflow.execution_graph,
+    )
+    steps_to_discard = set()
+    next_steps = execution_coordinator.get_steps_to_execute_next(
+        steps_to_discard=steps_to_discard
+    )
+    while next_steps is not None:
+        steps_to_discard = await execute_steps(
+            next_steps=next_steps,
+            workflow=workflow,
+            runtime_parameters=runtime_parameters,
+            execution_cache=execution_cache,
+            max_concurrent_steps=max_concurrent_steps,
+            step_execution_mode=step_execution_mode,
+        )
+        next_steps = execution_coordinator.get_steps_to_execute_next(
+            steps_to_discard=steps_to_discard
+        )
+    return construct_workflow_output(
+        workflow_outputs=workflow.workflow_definition.outputs,
+        execution_cache=execution_cache,
+    )
+
+
+async def execute_steps(
+    next_steps: List[str],
+    workflow: CompiledWorkflow,
+    runtime_parameters: Dict[str, Any],
+    execution_cache: ExecutionCache,
+    max_concurrent_steps: int,
+    step_execution_mode: StepExecutionMode,
+) -> Set[str]:
+    logger.info(f"Executing steps: {next_steps}. Execution mode: {step_execution_mode}")
+    nodes_to_discard = set()
+    steps_batches = list(
+        make_batches(iterable=next_steps, batch_size=max_concurrent_steps)
+    )
+    for steps_batch in steps_batches:
+        logger.info(f"Steps batch: {steps_batch}")
+        coroutines = [
+            safe_execute_step(
+                step=step,
+                workflow=workflow,
+                runtime_parameters=runtime_parameters,
+                execution_cache=execution_cache,
+                step_execution_mode=step_execution_mode,
+            )
+            for step in steps_batch
+        ]
+        results = await asyncio.gather(*coroutines)
```

Comment:
you are right in terms of observation and you are also right regarding possible optimisation.

That can be implemented via adding one dimension for `StepExecutionCoordinator` - which would decide on
parallel branching [dim_1] and max-depth-wise graph execution in branch (instead of single step) - [dim_2]

@grzegorz-roboflow - I propose to introduce that change in next PR, as this is performance improvement, not bug for in this EE. That also should not impact blocks API